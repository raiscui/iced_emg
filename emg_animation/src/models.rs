pub mod color;
pub mod opacity;
use std::{rc::Rc, time::Duration};

use im::Vector;

use crate::Debuggable;

/*
 * @Author: Rais
 * @Date: 2021-05-07 19:19:51
 * @LastEditTime: 2021-05-12 19:32:47
 * @LastEditors: Rais
 * @Description:
 */
// 1秒(s) ＝1000毫秒(ms)
// 1毫秒(ms)＝1000微秒(us)
// 1微秒 micro (us)＝1000纳秒(ns)
#[derive(Clone, Debug)]
pub(crate) enum Interpolation {
    Spring {
        stiffness: f64,
        damping: f64,
    },
    Easing {
        progress: f64,
        duration: Duration,
        start: f64,
        ease: Rc<Debuggable<dyn Fn(f64) -> f64>>,
    },
    AtSpeed {
        per_second: f64,
    },
}

#[derive(Clone, Debug)]
pub struct Motion {
    pub(crate) position: f64,
    pub(crate) velocity: f64,
    pub(crate) target: f64,
    pub(crate) interpolation: Interpolation,
    pub(crate) unit: String,
    pub(crate) interpolation_override: Option<Interpolation>,
}

#[derive(Clone, Debug)]
pub struct CubicCurveMotion {
    control1: (Motion, Motion),
    control2: (Motion, Motion),
    point: (Motion, Motion),
}

#[derive(Debug, Clone)]
pub struct QuadraticCurveMotion {
    control: (Motion, Motion),
    point: (Motion, Motion),
}

#[derive(Debug, Clone)]
pub struct ArcMotion {
    x: Motion,
    y: Motion,
    radius: Motion,
    start_angle: Motion,
    end_angle: Motion,
}

#[derive(Clone, Debug)]
pub enum PathCommand {
    Move(Motion, Motion),
    MoveTo(Motion, Motion),
    Line(Motion, Motion),
    LineTo(Motion, Motion),
    Horizontal(Motion),
    HorizontalTo(Motion),
    Vertical(Motion),
    VerticalTo(Motion),
    Curve(CubicCurveMotion),
    CurveTo(CubicCurveMotion),
    Quadratic(QuadraticCurveMotion),
    QuadraticTo(QuadraticCurveMotion),
    SmoothQuadratic(Vec<(Motion, Motion)>),
    SmoothQuadraticTo(Vec<(Motion, Motion)>),
    Smooth(Vec<(Motion, Motion)>),
    SmoothTo(Vec<(Motion, Motion)>),
    ClockwiseArc(ArcMotion),
    AntiClockwiseArc(ArcMotion),
    Close,
}

#[derive(Clone, Debug)]
pub struct ShadowMotion {
    offset_x: Motion,
    offset_y: Motion,
    size: Motion,
    blur: Motion,
    red: Motion,
    green: Motion,
    blue: Motion,
    alpha: Motion,
}

#[derive(Clone, Debug)]
pub enum Property {
    Exact(String, String),
    Color(String, Motion, Motion, Motion, Motion),
    Shadow(String, bool, Box<ShadowMotion>),
    Prop(String, Motion),
    Prop2(String, Motion, Motion),
    Prop3(String, Motion, Motion, Motion),
    Prop4(String, Motion, Motion, Motion, Motion),
    Angle(String, Motion),
    Points(Vec<(Motion, Motion)>),
    Path(Vec<PathCommand>),
}
// propertyName : Property -> String
pub fn property_name(prop: &Property) -> &str {
    use Property::{Angle, Color, Exact, Path, Points, Prop, Prop2, Prop3, Prop4, Shadow};

    match prop {
        Exact(name, ..)
        | Color(name, ..)
        | Shadow(name, ..)
        | Prop(name, ..)
        | Prop2(name, ..)
        | Prop3(name, ..)
        | Prop4(name, ..)
        | Angle(name, ..) => name,

        Points(_) => "points",

        Path(_) => "path",
    }
}

#[allow(clippy::pub_enum_variant_names)]
#[derive(Clone, Debug)]
pub enum Step<Message> {
    _Step,
    To(Vec<Property>),
    ToWith(Vec<Property>),
    Set(Vec<Property>),
    Wait(Duration),
    Send(Message),
    Repeat(u32, Vec<Step<Message>>),
    Loop(Vec<Step<Message>>),
}

#[derive(Debug)]
pub struct Timing {
    pub(crate) current: Duration,
    pub(crate) dt: Duration,
}

#[derive(Debug)]
pub struct Animation<Message>
where
    Message: Clone,
{
    pub(crate) steps: Vec<Step<Message>>,
    pub(crate) style: Vec<Property>,
    pub(crate) timing: Timing,
    pub(crate) running: bool,
    pub(crate) interruption: Vector<(Duration, Vec<Step<Message>>)>,
}
#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Tick(pub Duration);

impl Tick {
    #[must_use]
    pub const fn subsec_millis(&self) -> u32 {
        self.0.subsec_millis()
    }
    #[must_use]
    pub fn new(millisecond: f64) -> Self {
        Self(Duration::from_micros(unsafe {
            (millisecond * 1000.).trunc().to_int_unchecked::<u64>()
        }))
    }
}

// mapPathMotion : (Motion -> Motion) -> PathCommand -> PathCommand
fn map_path_motion(func: &dyn Fn(Motion) -> Motion, cmd: PathCommand) -> PathCommand {
    use PathCommand::{
        AntiClockwiseArc, ClockwiseArc, Close, Curve, CurveTo, Horizontal, HorizontalTo, Line,
        LineTo, Move, MoveTo, Quadratic, QuadraticTo, Smooth, SmoothQuadratic, SmoothQuadraticTo,
        SmoothTo, Vertical, VerticalTo,
    };
    // let func_clone = func.clone();
    let map_coords = move |coords: Vec<(Motion, Motion)>| -> Vec<(Motion, Motion)> {
        // let func_clone2 = func_clone.clone();

        coords
            .into_iter()
            .map(|(x, y)| (func(x), func(y)))
            .collect()

        // List.map
        //     (\( x, y ) ->
        //         ( fn x
        //         , fn y
        //         )
        //     )
        //     coords
    };

    match cmd {
        Move(m1, m2) => Move(func(m1), func(m2)),

        MoveTo(m1, m2) => MoveTo(func(m1), func(m2)),

        Line(m1, m2) => Line(func(m1), func(m2)),

        LineTo(m1, m2) => LineTo(func(m1), func(m2)),

        Horizontal(motion) => Horizontal(func(motion)),

        HorizontalTo(motion) => HorizontalTo(func(motion)),

        Vertical(motion) => Vertical(func(motion)),

        VerticalTo(motion) => VerticalTo(func(motion)),

        Curve(CubicCurveMotion {
            control1,
            control2,
            point,
        }) => Curve(CubicCurveMotion {
            control1: (func(control1.0), func(control1.1)),
            control2: (func(control2.0), func(control2.1)),
            point: (func(point.0), func(point.1)),
        }),
        CurveTo(CubicCurveMotion {
            control1,
            control2,
            point,
        }) => CurveTo(CubicCurveMotion {
            control1: (func(control1.0), func(control1.1)),
            control2: (func(control2.0), func(control2.1)),
            point: (func(point.0), func(point.1)),
        }),
        Quadratic(QuadraticCurveMotion { control, point }) => Quadratic(QuadraticCurveMotion {
            control: (func(control.0), func(control.1)),
            point: (func(point.0), func(point.1)),
        }),
        QuadraticTo(QuadraticCurveMotion { control, point }) => QuadraticTo(QuadraticCurveMotion {
            control: (func(control.0), func(control.1)),
            point: (func(point.0), func(point.1)),
        }),
        SmoothQuadratic(coords) => SmoothQuadratic(map_coords(coords)),
        // SmoothQuadratic <| map_coords coords
        SmoothQuadraticTo(coords) => SmoothQuadraticTo(map_coords(coords)),
        Smooth(coords) => Smooth(map_coords(coords)),
        SmoothTo(coords) => SmoothTo(map_coords(coords)),
        ClockwiseArc(arc) => ClockwiseArc({
            let ArcMotion {
                x,
                y,
                radius,
                start_angle,
                end_angle,
            } = arc;

            ArcMotion {
                x: func(x),
                y: func(y),
                radius: func(radius),
                start_angle: func(start_angle),
                end_angle: func(end_angle),
            }
        }),
        AntiClockwiseArc(arc) => AntiClockwiseArc({
            let ArcMotion {
                x,
                y,
                radius,
                start_angle,
                end_angle,
            } = arc;

            ArcMotion {
                x: func(x),
                y: func(y),
                radius: func(radius),
                start_angle: func(start_angle),
                end_angle: func(end_angle),
            }
        }),
        Close => Close,
    }
}

// mapToMotion : (Motion -> Motion) -> Property -> Property
pub fn map_to_motion(func: &dyn Fn(Motion) -> Motion, prop: Property) -> Property {
    use Property::{Angle, Color, Exact, Path, Points, Prop, Prop2, Prop3, Prop4, Shadow};
    match prop {
        Exact(name, value) => Exact(name, value),

        Color(name, m1, m2, m3, m4) => Color(name, func(m1), func(m2), func(m3), func(m4)),

        Shadow(name, inset, shadow) => {
            let ShadowMotion {
                offset_x,
                offset_y,
                size,
                blur,
                red,
                green,
                blue,
                alpha,
            } = *shadow;

            Shadow(
                name,
                inset,
                Box::new(ShadowMotion {
                    offset_x: func(offset_x),
                    offset_y: func(offset_y),
                    size: func(size),
                    blur: func(blur),
                    red: func(red),
                    green: func(green),
                    blue: func(blue),
                    alpha: func(alpha),
                }),
            )
        }
        Prop(name, m1) => Prop(name, func(m1)),

        Prop2(name, m1, m2) => Prop2(name, func(m1), func(m2)),

        Prop3(name, m1, m2, m3) => Prop3(name, func(m1), func(m2), func(m3)),

        Prop4(name, m1, m2, m3, m4) => Prop4(name, func(m1), func(m2), func(m3), func(m4)),

        Angle(name, m1) => Angle(name, func(m1)),

        Points(ms) => Points(ms.into_iter().map(|(x, y)| (func(x), func(y))).collect()),
        // Points <|
        //     List.map
        //         (\( x, y ) =>
        //             ( func x
        //             , func y
        //             )
        //         )
        //         ms
        Path(cmds) => Path(
            cmds.into_iter()
                .map(|p_cmd| map_path_motion(func, p_cmd))
                .collect(),
        ), // Path <|
           //     List.map
           //         (mapPathMotion func)
           //         cmds
           //
    }
}
