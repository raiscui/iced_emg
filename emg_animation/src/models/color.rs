use emg_common::vector;
use seed_styles::Unit;

/*
 * @Author: Rais
 * @Date: 2021-05-10 15:31:40
 * @LastEditTime: 2023-01-13 16:48:11
 * @LastEditors: Rais
 * @Description:
 */
use crate::init_motion;

use super::Property;

pub struct Color {
    red: u8,
    green: u8,
    blue: u8,
    alpha: f64,
}

impl Color {
    #[must_use]
    pub const fn new(red: u8, green: u8, blue: u8, alpha: f64) -> Self {
        Self {
            red,
            green,
            blue,
            alpha,
        }
    }
}

fn custom_color(
    name: &str,
    Color {
        red,
        green,
        blue,
        alpha,
    }: Color,
) -> Property {
    Property::Color(
        name.into(),
        Box::new([
            init_motion(f64::from(red), Unit::Empty),
            init_motion(f64::from(green), Unit::Empty),
            init_motion(f64::from(blue), Unit::Empty),
            init_motion(alpha, Unit::Empty),
        ]),
    )
}
#[must_use]
pub fn fill(color: Color) -> Property {
    custom_color("fill", color)
}
