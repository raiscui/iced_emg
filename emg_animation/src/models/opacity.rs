use crate::custom;

/*
 * @Author: Rais
 * @Date: 2021-05-10 15:58:45
 * @LastEditTime: 2021-05-10 16:06:54
 * @LastEditors: Rais
 * @Description:
 */
// opacity : Float -> Animation.Model.Property
pub fn opacity(val: f64) -> super::Property {
    custom("opacity".to_string(), val, "".to_string())
}
