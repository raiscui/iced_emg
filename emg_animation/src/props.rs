use im_rc::Vector;
use tracing::error;

use crate::{func::list_find_dup, models::Property};

/*
 * @Author: Rais
 * @Date: 2021-06-02 12:40:24
 * @LastEditTime: 2021-08-29 22:48:19
 * @LastEditors: Rais
 * @Description:
 */

#[must_use]
pub fn is_transformation(prop: &Property) -> bool {
    // List.member (propertyName prop)
    [
        "rotate",
        "rotateX",
        "rotateY",
        "rotateZ",
        "rotate3d",
        "translate",
        "translate3d",
        "scale",
        "scale3d",
    ]
    .contains(&prop.name().as_str())
}

pub fn warn_for_double_listed_properties(props: &Vector<Property>) {
    let mut names = props
        .iter()
        .filter(|&prop| !is_transformation(prop))
        .map(Property::name)
        .collect::<Vec<_>>();
    names.sort_unstable();
    let dup = list_find_dup(|a, b| a == b, names.as_slice());
    if !dup.is_empty() {
        error!("{:?}", dup);
    }
    // ────────────────────────────────────────────────────────────────────────────────
}
