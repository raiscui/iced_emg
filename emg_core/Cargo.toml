[package]
name = "emg_core"
version = "0.1.0"
edition = "2018"
resolver = "2"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[profile.dev]
split-debuginfo = "unpacked"

[dependencies]
ordered-float = "2.1.1"
derive_more = "0.99.13"
