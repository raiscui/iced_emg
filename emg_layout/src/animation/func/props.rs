use emg_animation::func::list_find_dup;
use emg_animation::props::is_transformation;
use emg_state::{CloneStateVar, GStateStore};
use im::Vector;
use tracing::error;

use crate::animation::StateVarProperty;

pub fn warn_for_double_listed_properties(props: &Vector<StateVarProperty>) {
    let mut names = props
        .iter()
        .filter_map(|prop| {
            let p = prop.get();
            if is_transformation(&p) {
                None
            } else {
                Some(p.name().to_string())
            }
        })
        .collect::<Vec<_>>();
    names.sort_unstable();
    let dup = list_find_dup(|a, b| a == b, names.as_slice());
    if !dup.is_empty() {
        error!("{:?}", dup);
    }
    // ────────────────────────────────────────────────────────────────────────────────
}
