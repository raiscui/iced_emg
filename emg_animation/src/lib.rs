#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![warn(clippy::nursery)]
// ────────────────────────────────────────────────────────────────────────────────
#![feature(slice_concat_ext)]
#![feature(div_duration)]
#![feature(extend_one)]
// #![feature(array_map)]
#![feature(box_patterns)]
// #![feature(adt_const_params)]
#![feature(generic_const_exprs)]
#![feature(slice_take)]

pub mod func;
pub mod models;
pub mod props;
mod render;
use std::{collections::VecDeque, convert::TryInto};
// ────────────────────────────────────────────────────────────────────────────────
// use emg_debuggable::dbg4;
use std::{f64::consts::PI, fmt, rc::Rc, time::Duration};

use emg_core::{vector, SmallVec, Vector};
use models::{
    map_to_motion, map_to_motion_og, update_animation_og, Animation, AnimationOG, Interpolation,
    Property, Step, StepOG,
};
use models::{update_animation, PropertyOG};
use ordered_float::NotNan;
use props::{warn_for_double_listed_properties, warn_for_double_listed_properties_og};
use seed_styles::Unit;
pub const PROP_SIZE: usize = 3;
pub const STEP_SIZE: usize = 3;
pub const MOTION_SIZE: usize = 3;
// ────────────────────────────────────────────────────────────────────────────────
pub use crate::models::color::fill;
pub use crate::models::opacity::opacity;
pub use crate::models::opacity::opacity_og;
pub use crate::models::Tick;
pub use crate::models::Timing;
// ────────────────────────────────────────────────────────────────────────────────
use crate::models::{Easing, Motion};

#[allow(clippy::enum_glob_use)]
use crate::models::Interpolation::*;
// ────────────────────────────────────────────────────────────────────────────────

pub type Msg = models::Tick;
pub type AmStateOG<Message> = models::AnimationOG<Message>;
pub type AmState<Message> = models::Animation<Message>;

/// # Panics
///
/// Will panic if 'position' is NaN
pub fn init_motion(position: impl TryInto<NotNan<f64>>, unit: Unit) -> Motion {
    let p: NotNan<f64> = position.try_into().ok().unwrap();
    Motion {
        position: p,
        velocity: NotNan::default(),
        target: p,
        interpolation: Spring {
            stiffness: NotNan::new(170.).unwrap(),
            damping: NotNan::new(26.).unwrap(),
        },
        unit,
        interpolation_override: None,
    }
}

// initialState : List Animation.Model.Property -> Animation msg

#[must_use]
pub fn initial_state<Message>(current: SmallVec<[Property; PROP_SIZE]>) -> Animation<Message>
where
    Message: Clone,
{
    Animation {
        steps: VecDeque::new(),
        props: current,
        timing: Timing {
            current: Duration::ZERO,
            dt: Duration::ZERO,
        },
        running: false,
        interruption: vector![],
    }
}
#[must_use]
pub fn initial_state_og<Message>(current: Vector<PropertyOG>) -> AnimationOG<Message>
where
    Message: Clone,
{
    AnimationOG {
        steps: vector![],
        props: current,
        timing: Timing {
            current: Duration::ZERO,
            dt: Duration::ZERO,
        },
        running: false,
        interruption: vector![],
    }
}

// speed : { perSecond : Float } -> Animation.Model.Interpolation
fn speed(speed_value: f64) -> Interpolation {
    AtSpeed {
        per_second: NotNan::new(speed_value).unwrap(),
    }
}

/// # Panics
///
/// Will panic if number is nan
#[must_use]
pub fn default_interpolation_by_property(prop: &Property) -> Interpolation {
    use Property::{Angle, Color, Exact, Path, Points, Prop, Prop2, Prop3, Prop4, Shadow};
    // -- progress is set to 1 because it is changed to 0 when the animation actually starts
    // -- This is analogous to the spring starting at rest.
    let linear = |duration: Duration| {
        Easing(Easing {
            progress: NotNan::new(1.).unwrap(),
            start: NotNan::default(),
            duration,
            ease: Rc::new(dbg4!(Box::new(std::convert::identity::<f64>))),
        })
    };

    let default_spring = Spring {
        stiffness: NotNan::new(170.).unwrap(),
        damping: NotNan::new(26.).unwrap(),
    };

    match prop {
        Exact(..) | Shadow(..) | Prop(..) | Prop2(..) | Prop4(..) | Points(..) | Path(..) => {
            default_spring
        }

        Color(..) => linear(Duration::from_millis(400)),

        Prop3(name, ..) => {
            if name.as_str() == "rotate3d" {
                speed(PI)
            } else {
                default_spring
            }
        }

        Angle(_, _) => speed(PI),
    }
}
/// # Panics
///
/// Will panic if 'prop' is NaN
#[must_use]
pub fn default_interpolation_by_property_og(prop: &PropertyOG) -> Interpolation {
    use PropertyOG::{Angle, Color, Exact, Path, Points, Prop, Prop2, Prop3, Prop4, Shadow};
    // -- progress is set to 1 because it is changed to 0 when the animation actually starts
    // -- This is analogous to the spring starting at rest.
    let linear = |duration: Duration| {
        Easing(Easing {
            progress: NotNan::new(1.).unwrap(),
            start: NotNan::default(),
            duration,
            ease: Rc::new(dbg4!(Box::new(std::convert::identity::<f64>))),
        })
    };

    let default_spring = Spring {
        stiffness: NotNan::new(170.).unwrap(),
        damping: NotNan::new(26.).unwrap(),
    };

    match prop {
        Exact(..) | Shadow(..) | Prop(..) | Prop2(..) | Prop4(..) | Points(..) | Path(..) => {
            default_spring
        }

        Color(..) => linear(Duration::from_millis(400)),

        Prop3(name, ..) => {
            if name.as_str() == "rotate3d" {
                speed(PI)
            } else {
                default_spring
            }
        }

        Angle(_, _) => speed(PI),
    }
}

// setDefaultInterpolation : Animation.Model.Property -> Animation.Model.Property

#[must_use]
pub fn set_default_interpolation_og(prop: PropertyOG) -> PropertyOG {
    let i = default_interpolation_by_property_og(&prop);
    map_to_motion_og(
        &|mut m: Motion| -> Motion {
            m.interpolation = i.clone();
            m
        },
        prop,
    )
}

pub fn set_default_interpolation(prop: &mut Property) {
    let i = default_interpolation_by_property(prop);

    map_to_motion(
        |m: &mut Motion| {
            m.interpolation = i.clone();
        },
        prop,
    );
}

// style : List Animation.Model.Property -> Animation msg

#[must_use]
pub fn style<Message>(mut props: SmallVec<[Property; PROP_SIZE]>) -> Animation<Message>
where
    Message: Clone,
{
    //
    warn_for_double_listed_properties(&props);
    props.iter_mut().for_each(set_default_interpolation);
    initial_state(props)
}
#[must_use]
pub fn styleOG<Message>(props: Vector<PropertyOG>) -> AnimationOG<Message>
where
    Message: Clone,
{
    //
    warn_for_double_listed_properties_og(&props);
    initial_state_og(
        props
            .into_iter()
            .map(set_default_interpolation_og)
            .collect::<Vector<PropertyOG>>(),
    )
}

// ────────────────────────────────────────────────────────────────────────────────

///Sums all leading `Wait` steps and removes them from the animation.
///This is used because the wait at the start of an interruption works differently than a normal wait.

//    extractInitialWait : List (Animation.Model.Step msg) -> ( Time.Posix, List (Animation.Model.Step msg) )

#[must_use]
pub fn extract_initial_wait_og<Message>(
    steps: Vector<StepOG<Message>>,
) -> (Duration, Vector<StepOG<Message>>)
where
    Message: Clone,
{
    // case List.head steps of
    // Nothing ->
    //     ( Time.millisToPosix 0, [] )

    // Just step ->
    //     case step of
    //         Wait till ->
    //             let
    //                 ( additionalTime, remainingSteps ) =
    //                     extractInitialWait (List.drop 1 steps)
    //             in
    //             ( Time.millisToPosix (Time.posixToMillis till + Time.posixToMillis additionalTime), remainingSteps )

    //         _ ->
    //             ( Time.millisToPosix 0, steps )
    use StepOG::Wait;
    let front = steps.front().cloned();
    match front {
        None => (Duration::ZERO, steps),
        Some(step) => {
            if let Wait(till) = step {
                let (additional_time, remaining_steps) = extract_initial_wait_og(steps.skip(1));
                (till + additional_time, remaining_steps)
            } else {
                (Duration::ZERO, steps)
            }
        } // [step] => (Duration::ZERO, steps),
    }
}

/// # Panics
///
/// Never panic
#[must_use]
pub fn extract_initial_wait<Message>(
    //TODO check mut , opt all
    mut steps: VecDeque<Step<Message>>,
) -> (Duration, VecDeque<Step<Message>>)
where
    Message: Clone,
{
    // case List.head steps of
    // Nothing ->
    //     ( Time.millisToPosix 0, [] )

    // Just step ->
    //     case step of
    //         Wait till ->
    //             let
    //                 ( additionalTime, remainingSteps ) =
    //                     extractInitialWait (List.drop 1 steps)
    //             in
    //             ( Time.millisToPosix (Time.posixToMillis till + Time.posixToMillis additionalTime), remainingSteps )

    //         _ ->
    //             ( Time.millisToPosix 0, steps )
    use Step::Wait;
    let front = steps.front();
    if let Some(Wait(_)) = front {
        let till = steps
            .pop_front()
            .and_then(|x| x.try_into_wait().ok())
            .unwrap();
        let (additional_time, remaining_steps) = extract_initial_wait(steps);
        (till + additional_time, remaining_steps)
    } else {
        (Duration::ZERO, steps)
    }
}

///Interrupt any running animations with the following animation.
// interrupt : List (Animation.Model.Step msg) -> Animation msg -> Animation msg

pub fn interrupt_og<Message>(
    steps: Vector<StepOG<Message>>,
    model: &mut AnimationOG<Message>,
) -> &mut AnimationOG<Message>
where
    Message: Clone,
{
    model
        .interruption
        .push_front(extract_initial_wait_og(steps));
    model.running = true;
    model
}
pub fn interrupt<Message>(
    steps: impl Into<VecDeque<Step<Message>>>,
    model: &mut Animation<Message>,
) -> &mut Animation<Message>
where
    Message: Clone,
{
    model
        .interruption
        .push_front(extract_initial_wait(steps.into()));
    model.running = true;
    model
}
pub fn replace<Message>(
    steps: impl Into<VecDeque<Step<Message>>>,
    model: &mut Animation<Message>,
) -> &mut Animation<Message>
where
    Message: Clone,
{
    model.interruption = vector![extract_initial_wait(steps.into())];
    model.running = true;
    model
}
pub fn replace_og<Message>(
    steps: Vector<StepOG<Message>>,
    model: &mut AnimationOG<Message>,
) -> &mut AnimationOG<Message>
where
    Message: Clone,
{
    model.interruption = vector![extract_initial_wait_og(steps)];
    model.running = true;
    model
}

// {-| Repeat a number of steps until interrupted.
//     -}
//     loop : List (Animation.Model.Step msg) -> Animation.Model.Step msg
//     loop steps =
//         Loop steps
pub fn loop_am<Message>(steps: impl Into<VecDeque<Step<Message>>>) -> Step<Message>
where
    Message: Clone,
{
    Step::Loop(steps.into())
}
#[must_use]
pub fn loop_am_og<Message>(steps: Vector<StepOG<Message>>) -> StepOG<Message>
where
    Message: Clone,
{
    StepOG::Loop(steps)
}

#[must_use]
pub fn to<Message>(props: SmallVec<[Property; PROP_SIZE]>) -> Step<Message>
where
    Message: Clone,
{
    Step::To(props)
}

#[must_use]
pub fn to_og<Message>(props: Vector<PropertyOG>) -> StepOG<Message>
where
    Message: Clone,
{
    StepOG::To(props)
}

// custom : String -> Float -> String -> Animation.Model.Property
fn custom(name: &str, value: f64, unit: Unit) -> Property {
    Property::Prop(name.into(), init_motion(value, unit))
}
fn custom_og(name: &str, value: f64, unit: Unit) -> PropertyOG {
    PropertyOG::Prop(name.into(), init_motion(value, unit))
}

/// Update an animation.
pub fn update<Message: std::clone::Clone + std::fmt::Debug>(
    tick: Tick,
    animation: &mut Animation<Message>,
) {
    update_animation(tick, animation);
}
pub fn update_og<Message: std::clone::Clone + std::fmt::Debug>(
    tick: Tick,
    animation: &mut AnimationOG<Message>,
) {
    update_animation_og(tick, animation);
}
// ────────────────────────────────────────────────────────────────────────────────

// ────────────────────────────────────────────────────────────────────────────────
#[derive(Clone)]
pub struct Debuggable<T> {
    text: &'static str,
    value: T,
}
impl<T> Eq for Debuggable<T> {}
impl<T> PartialEq for Debuggable<T> {
    fn eq(&self, other: &Self) -> bool {
        self.text.eq(other.text)
    }
}

impl<T> std::ops::Deref for Debuggable<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

/// Produce a Debuggable<T> from an expression for T
#[macro_export]
macro_rules! dbg4 {
    ($($body:tt)+) => {
        Debuggable {
            text: stringify!($($body)+),
            value: $($body)+,
        }
    };
}

// Note: this type is unsized

impl<T> fmt::Debug for Debuggable<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.text)
    }
}
// ────────────────────────────────────────────────────────────────────────────────

#[allow(dead_code)]
#[cfg(test)]
mod tests {
    use std::time::Duration;

    use emg_core::vector;

    use crate::{
        extract_initial_wait_og, fill, interrupt_og,
        models::{color::Color, opacity::opacity_og, update_animation_og, StepOG},
        styleOG, to_og, AmStateOG, Tick,
    };

    #[derive(Clone, Debug)]
    enum Message {
        A,
        B,
    }
    #[test]
    fn it_works() {
        let styles: AmStateOG<Message> = styleOG(vector![fill(Color::new(0, 0, 0, 1.))]);
        println!("{:#?}", styles);
    }
    #[test]
    fn test_extract_initial_wait() {
        let xx = vector![
            StepOG::Wait(Duration::from_millis(16)),
            StepOG::_Step,
            StepOG::Send(Message::A),
        ];
        println!("{:#?}", xx);
        let ff = extract_initial_wait_og(vector![
            StepOG::Wait(Duration::from_millis(16)),
            StepOG::_Step,
            StepOG::Send(Message::A),
        ]);
        println!("{:#?}", &ff);
        let v = (
            Duration::from_millis(16),
            vec![StepOG::_Step, StepOG::Send(Message::A)],
        );
        assert_eq!(format!("{:?}", v), format!("{:?}", ff))
    }
    #[test]
    fn test_update_animation() {
        let mut am_state: AmStateOG<Message> = styleOG(vector![opacity_og(1.)]);
        insta::assert_debug_snapshot!("init", &am_state);

        interrupt_og(
            vector![
                to_og(vector![opacity_og(0.)]),
                to_og(vector![opacity_og(1.)])
            ],
            &mut am_state,
        );
        insta::assert_debug_snapshot!("interrupt", &am_state);

        let mut now = Duration::from_millis(10000);
        update_animation_og(Tick(now), &mut am_state);
        insta::assert_debug_snapshot!("am1-first", &am_state);

        now += Duration::from_millis(16);
        update_animation_og(Tick(now), &mut am_state);
        insta::assert_debug_snapshot!("am2", &am_state);

        now += Duration::from_millis(17);
        update_animation_og(Tick(now), &mut am_state);
        insta::assert_debug_snapshot!("am3", &am_state);

        for _ in 0..180 {
            now += Duration::from_millis(17);
            update_animation_og(Tick(now), &mut am_state);
        }
        println!("{:#?}", &am_state);
        insta::assert_debug_snapshot!("am_last", &am_state);
    }
    #[test]
    fn test_interrupt() {
        let mut am_state: AmStateOG<Message> = styleOG(vector![opacity_og(1.)]);
        let interrupt1 = interrupt_og(
            vector![
                to_og(vector![opacity_og(0.)]),
                to_og(vector![opacity_og(1.)])
            ],
            &mut am_state,
        );
        println!("{:#?}", interrupt1);
    }
}
