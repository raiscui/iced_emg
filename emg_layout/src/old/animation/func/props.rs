use emg_animation::func::list_find_dup;
use emg_animation::props::is_transformation_og;
use emg_core::Vector;
use emg_state::CloneStateVar;
use tracing::error;

use crate::old::animation::StateVarPropertyOG;

pub fn warn_for_double_listed_properties(props: &Vector<StateVarPropertyOG>) {
    let mut names = props
        .iter()
        .filter_map(|prop| {
            let p = prop.get();
            if is_transformation_og(&p) {
                None
            } else {
                Some(p.name())
            }
        })
        .collect::<Vec<_>>();
    names.sort_unstable();
    let dup = list_find_dup(|a, b| a == b, names.as_slice());
    if !dup.is_empty() {
        error!("{:?}", dup);
    }
    // ────────────────────────────────────────────────────────────────────────────────
}
