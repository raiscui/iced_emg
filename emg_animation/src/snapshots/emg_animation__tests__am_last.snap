---
source: emg_animation/src/lib.rs
expression: "&am_state"

---
Animation {
    steps: [],
    props: [
        Prop(
            "opacity",
            Motion {
                position: NotNan(
                    1.0,
                ),
                velocity: NotNan(
                    0.0,
                ),
                target: NotNan(
                    1.0,
                ),
                interpolation: Spring {
                    stiffness: NotNan(
                        170.0,
                    ),
                    damping: NotNan(
                        26.0,
                    ),
                },
                unit: None,
                interpolation_override: None,
            },
        ),
    ],
    timing: Timing {
        current: 13.093s,
        dt: 17ms,
    },
    running: false,
    interruption: [],
}
